_posts/2020-12-09-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,528 @@
+```
+
+We have uncovered another edge case! The balancer token disallows transfers to the zero address. If
+you are interested in learning about more ERC20 edge cases an extensive list is maintained at
+[weird-er20](https://github.com/xwvvvvwx/weird-erc20).
+
+## Interactive Exploration
+
+`hevm` also includes a visual debugger, and we can use this to interactively explore the execution
+tree. You can enter the debugger by running `dapp debug` from the root of your `dapp project`. You
+will see a list of test methods, and once you select one you will be dropped into an interactive
+debugging session.
+
+You can press `h` to bring up a help view, `n` to step forwards, and `p` to step back. If you press
+`e` in symbolic test you will jump to the next branching point. Once there you can press `0` to choose
+the branch which does not jump, and `1` to choose the branch that does.
+
+Note that the interactive debugger will also function when executing against mainnet state.
+
+A small demonstration video can be found below:
+
+[![asciicast](https://asciinema.org/a/z1gg559jjOFnmfwI8A5LxTM0T.svg)](https://asciinema.org/a/z1gg559jjOFnmfwI8A5LxTM0T)
+
+## Limitations, Assumptions & Future Work
+
+#### Non-Linearity (`safeMul`)
+
+The symbolic execution engine in `hevm` is backed by an [SMT
+solver](https://en.wikipedia.org/wiki/Satisfiability_Modulo_Theories) (currently either `z3` or
+`cvc4` are supported). Expressions involving non-linear arithmetic (multiplication, division or
+exponentiation by symbolic variables) are extremely challenging or impossible for SMT solvers, and
+it will often not be practical to symbolically execute tests involving lots of non-linear
+arithmetic.
+
+Unfortunately, non-linear arithmetic is quite common in real world contracts (e.g.
+[`safeMul`](https://github.com/dapphub/ds-math/blob/master/src/math.sol#L25) can easily
+involve both a multiplication and a division by a symbolic variable).
+
+We hope to include optimizations in future releases of `hevm` that reduce the load on the solver
+when executing contracts that make use of common non-linearities (`safeMul` included).
+
+#### Symbolic Representation of Dynamic Types
+
+`hevm` is currently unable to represent dynamic types (e.g. `bytes`, `string`) symbolically. Tests
+that need symbolic representations of dynamic data will currently fail with an `Unsupported symbolic
+abiencoding` error.
+
+We intend to lift this restriction in a future release of `hevm`.
+
+#### Symbolic Constructor Arguments
+
+Contract bytecode is currently assumed by `hevm` to be completely concrete. As constructor arguments
+are implemented on the EVM level by appending data to the contract's `creationCode`, symbolic
+execution of contract constructors where the arguments are set to symbolic values will currently
+fail with an `UnexpectedSymbolicArg` error.
+
+We intend to lift this restriction in a future release of `hevm`.
+
+#### State Explosion
```

Comment:
mention loops and `--max-iterations`?

Reply:
added sections for loops and unknown calls.